pkg/controller/bucket/bucket_controller.go

```diff
@@ -0,0 +1,200 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bucket
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	utilversion "k8s.io/apimachinery/pkg/util/version"
+
+	kubeclientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/retry"
+	"k8s.io/client-go/util/workqueue"
+
+	"github.com/kubernetes-sigs/container-object-storage-interface-api/apis/objectstorage.k8s.io/v1alpha1"
+	bucketclientset "github.com/kubernetes-sigs/container-object-storage-interface-api/clientset"
+	"github.com/kubernetes-sigs/container-object-storage-interface-api/controller"
+
+	osspec "github.com/kubernetes-sigs/container-object-storage-interface-spec"
+
+	"k8s.io/klog"
+
+	"golang.org/x/time/rate"
+)
+
+// bucketListener manages Bucket objects
+type bucketListener struct {
+	kubeClient        kubeclientset.Interface
+	bucketClient      bucketclientset.Interface
+	provisionerClient osspec.ProvisionerClient
+
+	// The name of the provisioner for which this controller dynamically
+	// provisions buckets.
+	provisionerName string
+	kubeVersion     *utilversion.Version
+}
+
+// NewBucketController returns a controller that manages Bucket objects
+func NewBucketController(provisionerName string, client osspec.ProvisionerClient) (*controller.ObjectStorageController, error) {
+	rateLimit := workqueue.NewMaxOfRateLimiter(
+		workqueue.NewItemExponentialFailureRateLimiter(5*time.Second, 60*time.Minute),
+		&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
+	)
+
+	identity := fmt.Sprintf("object-storage-sidecar-%s", provisionerName)
+	bc, err := controller.NewObjectStorageController(identity, "bucket-controller", 5, rateLimit)
+	if err != nil {
+		return nil, err
+	}
+
+	bl := bucketListener{
+		provisionerName:   provisionerName,
+		provisionerClient: client,
+	}
+	bc.AddBucketListener(&bl)
+
+	return bc, nil
+}
+
+// InitializeKubeClient initializes the kubernetes client
+func (bl *bucketListener) InitializeKubeClient(k kubeclientset.Interface) {
+	bl.kubeClient = k
+
+	serverVersion, err := k.Discovery().ServerVersion()
+	if err != nil {
+		klog.Errorf("unable to get server version: %v", err)
+	} else {
+		bl.kubeVersion = utilversion.MustParseSemantic(serverVersion.GitVersion)
+	}
+}
+
+// InitializeBucketClient initializes the object storage bucket client
+func (bl *bucketListener) InitializeBucketClient(bc bucketclientset.Interface) {
+	bl.bucketClient = bc
+}
+
+// Add will call the provisioner and add a bucket
+func (bl *bucketListener) Add(ctx context.Context, obj *v1alpha1.Bucket) error {
+	klog.V(1).Infof("bucketListener: add called for bucket %s", obj.Name)
+
+	// Verify this bucket is for this provisioner
+	if !strings.EqualFold(obj.Spec.Provisioner, bl.provisionerName) {
```

Comment:
who provides the provisionerName for the Sidecar?

Reply:
The sidecar queries the driver for that information.  It's in code to be merged after this is in.